sandbox/apparmor/apparmor.go

```diff
@@ -412,3 +412,14 @@ func MockFeatures(kernelFeatures []string, kernelError error, parserFeatures []s
 	}
 
 }
+
+// ProfileNameForPath returns he name of the profile for an executable at given
+// path.
+func ProfileNameForPath(path string) (string, error) {
+	path = filepath.Clean(path)
+	relToRoot, err := filepath.Rel("/", path)
+	if err != nil {
+		return "", fmt.Errorf("cannot make path %q relative to /", path)
+	}
+	return strings.Replace(relToRoot, "/", ".", -1), nil
```

Comment:
While this is an ok guess, note that the use of '.' is a convention (and a deprecated one) and not something to rely upon. Ultimately, this is a packaging issue and if a distro is choosing not to use a particular system profile location, then IMHO we must have a may to tell snapd at build time where that location is.

We've historically shipped the snap-confine (and snap-confine.real) profiles as system profiles in /etc/apparmor.d since apparmor is expected to start before snaps and this allows for the snap-confine profile to be loaded in time for snaps to start. Because it is a system profile, it allows distro packagers to use a different name, etc which if I'm reading this PR correctly, is the source of the problem. There are a couple of ways to solve this (perhaps more):
1. make the path to the system profile build time configurable (instead of trying to guess)
2. stop shipping the profile as a system profile and manage this wholly within snapd

For some time for the reexec scenario we've used profiles in `/var/lib/snapd/apparmor/profiles/snap-confine....`  and had snapd manage them. We also have had for some time snapd.apparmor.service and most things in packaging/ (all?) use it these days. I suggest pursuing '2' and go the last step: stop shipping a system profile in /etc/apparmor.d, start shipping the 'supplied by the deb/rpm/etc' file somewhere controlled by snapd (eg, /usr/lib/snapd/snap-confine.apparmor or /var/lib/snapd/apparmor/profiles/snap-confine.distro), adjust snapd.apparmor.service to also load this profile, adjust this code to load this profile location.

There might be additional steps and you'll want to make sure all packaging is using snapd.apparmor.service (Ubuntu for example historically did not), but the general idea is don't ship a system profile, have snapd manage it and therefore avoid any ambiguity on the name.